### DISCUSSION TITLE
Snaps Wishlist

### DISCUSSION THREAD

**Discussion by Montoya** [MetaMask/snaps#474] [Ideas]
The following are types of Snaps that we would love to see from the community. *Note: some of these may require the Snaps Platform team to ship more APIs before they are possible.*

- Smart Contract Account / multisig account support
- Account abstraction features (session keys, tx batching, etc)
- Transaction insights / simulations
- Known address / contract management, phishing prevention
- Messaging and notification services (like Waku, EPNS)
- Decentralized identity / privacy-enabled identity
- DAO tooling like notifications for vote proposals, show percentage of ownership
- Price feed data for tokens, NFTs, fractional ownership
- Track staked assets, yield generating assets
- L2 support including ZK networks and rollups
- Protocol snaps (like Avalanche, Tezos)
- Hardware wallet snaps (for hardware wallets not already supported by MetaMask)
- Key backup (like storing fragments of your private key in cloud based services to make key recovery easier)

If you are working on one of these use cases, or need support from the Snaps Platform team to enable one of these use cases, let us know!

**Comment by HarishRocks** [comment]
Hey @Montoya I'll be exploring how to build hardware wallet snaps for our Hardware wallet [Cypherock](https://cypherock.com/)
As this could be a need for the community of upcoming hardware wallets, I would be glad to work on a template-snap for hardware wallet that could be consumed easily by other providers. 
It would be great if the team could share any kind of information already researched particular for this use-case 
Thanks

**Comment by cryptoKevinL** [comment]
Hi @Montoya - we have wanted to integrate our wallet-to-wallet chat protocol with MM for some time, just found out about snaps.  Is it best to just do our best to integrate using our own UI for now since we cannot make UI changes currently?  Open to a conversation if that helps the use cases you are looking for in messaging/notifications.

**Reply by Montoya** [reply]
Yes for now building your own UI into a dApp is the best approach until we enable more UI customization in the Snaps API. I will be sharing more information in this board about UI API features as they become available.

**Comment by andyv09** [comment]
Hey guys!

We are a team working on [SSI Snap](https://github.com/blockchain-lab-um/ssi-snap), a MetaMask Snap that enables support for management of DIDs, storage of Verifiable Credentials, and generation of Verifiable Presentations. We've made quite some progress on the SSI Snap since the [first article](https://medium.com/@blockchainlabum/its-time-to-prove-your-worth-in-dao-ssi-using-metamask-snaps-part-1-3-cfed7c10e6f7). 

Feel free to check out [part 2](https://medium.com/@blockchainlabum/its-time-to-prove-your-worth-in-dao-ssi-using-metamask-snaps-part-2-3-17eb98678054) of our series, where we discuss our progress and showcase a demo using Snapshot!

We've also open-sourced the code and made public docs, more can be found [here](https://medium.com/@blockchainlabum/open-sourcing-ssi-snap-for-metamask-aaa176775be2). 

We'd love to hear your thoughts!

**Comment by gulshanvas** [comment]
Hi @Montoya ,
I am working on Snap to provide seamless Messaging and notification service expertise.
Are there specific any JSON-RPC methods available for it ?

**Reply by ritave** [reply]
Hello @gulshanvas 

Could you expand on the concept of "notification service"?
We have [`snap_notify`](https://docs.metamask.io/guide/snaps-rpc-api.html#snap-notify) that can send both native system notifications and MetaMask notifications.

The messaging part is up to you, you'll probably need [`endowment:internet-access`](https://docs.metamask.io/guide/snaps-permissions.html#endowment-network-access)

If those two methods are not enough, we'll happy to listen for new feature requests.

**Reply by gulshanvas** [reply]
Hi @ritave ,
I am planning to build out snap which would help manage **notification service** like EPNS from metamask itself. Basically subscribing, unsubscribing actions of specific channels of EPNS from metamask window.

Does metamask has discord server or telegram ?

**Reply by ritave** [reply]
@gulshanvas 
Consensys has a [Discord server](https://discord.gg/consensys). Though Snaps developers mostly monitor this Discussion board in general rather than the Discord.

**Reply by Montoya** [reply]
There is also a cron job API coming soon to Flask that can be used to check for new notifications / messages on a regular interval. 

Currently there is no way to display messages within MetaMask, that will require custom UI which is not ready yet. But right now you can display notifications and then surface the messages within a dapp.

**Reply by gulshanvas** [reply]
Hi @Montoya ,
Will be eagerly waiting for cron job API.
Thanks for the update.

**Reply by gulshanvas** [reply]
Hi @ritave @Montoya ,
I would like to know what's journey after a snap is build out till distribution to end users.
How the user would get to know about snap and process of installation of snap ? Basically inclusion of features provided snap into metamask.

**Reply by Montoya** [reply]
After publishing your snap to NPM, you should publish a dapp to the web and invite people to visit your dapp to install your snap. This dapp can include information about your snap and why people should install it. This is the primary distribution method. 

When we enable Snaps in the stable extension, we will include announcements and user guides about snaps so they can understand how snaps work and what to expect when installing a snap.

**Reply by gulshanvas** [reply]
Hi @ritave @Montoya ,
At ETHIndia 2022, built a snap (PushSnap) for Push protocol to receive notifications directly into metamask browser extension.
Is their any central place where it can get listed ?

**Reply by Montoya** [reply]
Hi @gulshanvas, right now there is no official directory for snaps. I recommend making a dapp so people can go to your website and install the snap. Then, you can make a new topic on this discussion board so others can try it. 

Right now Snaps are only live in Flask which is a developer preview version of MetaMask. So mainly it will be other developers who will see it. When Snaps goes live in the stable extension next year then there will probably be other ways to publicize snaps, but we do not have any plans to make an official "directory."

**Comment by lljxx1** [comment]
Hi, Is possible to include the page domain in `onTransaction `?

**Reply by ritave** [reply]
Not yet, but it's been in talks internally and I've created an issue for it #956.

**Reply by lljxx1** [reply]
Thanks!

**Comment by artk42** [comment]
Hi @Montoya,

We consider focusing on the last bullet specifically:
- Key backup (like storing fragments of your private key in cloud-based services to make key recovery easier))

Since it's what we are best at (sharding with strong authentication and fault-tolerance).
However, I believe this one point is also one of those that needs "ship more APIs before they are possible". So the question is how is it possible to communicate about the current state of development (only get-bip32 and -bip44 methods doesn't seem to help deal with seed sharding) and upcoming details?
I think we might have some suggestions on how it could be implemented in the most flexible way.

**Reply by ritave** [reply]
Hey @artk42, if you have any suggestions on new APIs we'd love to hear them.

We run [Snaps Improvement Proposals](https://github.com/MetaMask/SIPs) repository where we work with the community on new APIs using a technical process. We'd be happy to have your contributions there.

**Comment by shipunyc** [comment]
I'm looking for "Smart Contract Account / multisig account support" that allow DApp to trigger a tx on a smart contract account.

**Comment by Arch0125** [comment]
Hey @Montoya, already working on first four points and here's the demo [SafeSnap](https://devfolio.co/projects/snapsafe-6b93)
Currently working key recovery using zk-mpc where fragments are decided by Shaamir's Secret sharing algorithm and given tpo trusted wallets/SC Accounts instead of cloud servers
